loginserver/client.cpp

```diff
@@ -583,6 +585,70 @@ void Client::DoSuccessfulLogin(
 	m_client_status = cs_logged_in;
 }
 
+void Client::SendExpansionPacketData(PlayerLoginReply_Struct& plrs)
+{
+	SerializeBuffer buf;
+	//from eqlsstr_us.txt id of each expansion, excluding 'Everquest'
+	int ExpansionLookup[20] = { 3007, 3008, 3009, 3010,	3012,
+								3014, 3031, 3033, 3036, 3040,
+								3045, 3046, 3047, 3514, 3516,
+								3518, 3520, 3522, 3524 };
+
+
+	//Get the active expansion from the database Rule:WorldExpansionSettings.  Do this only once.  Requires restart to effect change
+	if (server.options.IsDisplayExpansions()) {
+
+		
+		int32_t expansion = server.options.GetMaxExpansions();
+		int32_t owned_expansion = (expansion << 1) | 1;
+
+		if (m_client_version == cv_sod) {
+
+			// header info of packet.  Requires OP_LoginExpansionPacketData=0x0031 to be in login_opcodes_sod.conf
+			buf.WriteInt32(0x00);
+			buf.WriteInt32(0x01);
+			buf.WriteInt16(0x00);
+			buf.WriteInt32(19); //number of expansions to include in packet
+
+			//generate expansion data
+			for (int i = 0; i < 19; i++)
+			{
+				buf.WriteInt32(i);													//sequenctial number
+				buf.WriteInt32((expansion & (1 << i)) == (1 << i) ? 0x01 : 0x00);	//1 own 0 not own
+				buf.WriteInt8(0x00);
+				buf.WriteInt32(ExpansionLookup[i]);									//from eqlsstr_us.txt
+				buf.WriteInt32(0x179E);												//from eqlsstr_us.txt for buttons/order
+				buf.WriteInt32(0xFFFFFFFF);											//end identification
+				buf.WriteInt8(0x0);													//force order window to appear 1 appear 0 not appear
+				buf.WriteInt8(0x0);
+				buf.WriteInt32(0x0000);
+				buf.WriteInt32(0x0000);
+				buf.WriteInt32(0xFFFFFFFF);
+			}
+
+			auto out = std::make_unique<EQApplicationPacket>(OP_LoginExpansionPacketData, buf);
+			m_connection->QueuePacket(out.get());
+
+		}
+		else if (m_client_version == cv_titanium)
+		{
+			if (expansion >= EQ::expansions::bitPoR)
+			{
+				// Titanium shipped with 10 expansions.  Set owned expansions to be max 10.
+				plrs.offer_min_days = ((EQ::expansions::bitDoD << 2) | 1) - 2;
```

Comment:
Sorry I didn't notice this before but the expansions enum also has masks that you can set this directly to.

Since titanium has a max of 10 expansions should it be `EQ::expansions::maskDoD`  (1023)? or am I misunderstanding why it's also including bitPoR (if so this can be set directly to `maskPoR`)

Reply:
My logic is based on the fact that you can force Titanium to display beyond its limit of 10.  For example, You own 18 of 27 expansions.  However, the equivalent strings are not in the eqls file and therefore the english does not work.  Therefore thought it best to add the limit.  Further, I wanted to use the same masking throughout for consistency and then create the formula to populate the bits to the right of the max.  Hope that makes sense.  :-)